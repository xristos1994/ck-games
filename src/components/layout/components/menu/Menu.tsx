import React, { FC, ReactElement, useLayoutEffect } from 'react';
import { Link } from 'gatsby';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { classnames, compose } from '@utils/component-utils';
import { withTranslation, ITranslate } from '@models/i18n/hoc';
import { Button } from '@components';
import { isMenuOpen } from '@models/layout/props';
import { selectedGame } from '@models/website/props';
import { setIsMenuOpen } from '@models/layout/actions';
import { availableLangs, lang } from '@models/i18n/props';
// import { availableLangs as _availableLangs } from '@models/i18n//utils';
import { setLang } from '@models/i18n/actions';
import { IState } from '@models/interfaces';
import { ArrowUpIcon, ArrowDownIcon } from '@components/icons';
import { AvailableGames } from '@models/website/interfaces';
import { LangFlagMap } from '@models/i18n/utils';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require('./Menu.module.css');

interface IProps {
  selectedGame: string | null;
  isMenuOpen: boolean;
  setIsMenuOpen: (isMenuOpen: boolean) => void;
  availableLangs: { code: string; label: string }[];
  lang: { code: string; label: string };
  setLang: (lang: IProps['lang']) => void;
  t: ITranslate;
}

export const _Menu: FC<IProps> = ({
  isMenuOpen,
  setIsMenuOpen,
  selectedGame,
  availableLangs,
  lang,
  setLang,
  t
}): ReactElement => {
  const isSSR = typeof window === 'undefined';

  const h1Title = (
    <h1 className={styles.menuTitle}>
      {isSSR && !selectedGame
        ? t('Menu Title SSR')
        : (isMenuOpen || isSSR) && selectedGame === AvailableGames.pantomime
          ? t('Pantomime')
          : (isMenuOpen || isSSR) && selectedGame === AvailableGames.tikTakBoom
            ? t('Tik-Tak-Boom')
            : isMenuOpen && !selectedGame
              ? t('CK-Games')
              : null}
    </h1>
  );

  const pantomimeDescription = (
    <>
      <div className={styles.gameDescription}>
        {isSSR || (!selectedGame && isMenuOpen) ? <h2>{t('Pantomime')}</h2> : null}
      </div>
      <div className={styles.gameDescription} dangerouslySetInnerHTML={{ __html: t('Pantomime Description HTML') }} />
    </>
  );

  const tikTakBoomDescription = (
    <>
      <div className={styles.gameDescription}>
        {isSSR || (!selectedGame && isMenuOpen) ? <h2>{t('Tik-Tak-Boom')}</h2> : null}
      </div>
      <div
        className={styles.gameDescription}
        dangerouslySetInnerHTML={{ __html: t('Tik-Tak-Boom Description HTML') }}
      />
    </>
  );

  const gameDescription = (
    <>
      {isSSR || (isMenuOpen && (selectedGame === AvailableGames.tikTakBoom || !selectedGame))
        ? tikTakBoomDescription
        : null}
      {isMenuOpen && !selectedGame ? <hr /> : null}
      {isSSR || (isMenuOpen && (selectedGame === AvailableGames.pantomime || !selectedGame))
        ? pantomimeDescription
        : null}
    </>
  );

  const homeButton = isMenuOpen && !!selectedGame && (
    <Link to={'/'}>
      <Button className={classnames(styles.homeButton)}>{t('Home')}</Button>
    </Link>
  );

  return (
    <div
      className={classnames(styles.menuContainer, {
        [styles.openMenu]: isMenuOpen
      })}
    >
      {
        (isMenuOpen || !selectedGame) ? (
          <div className={styles.langSelector}>
            {availableLangs.map((_lang) => {
              const isSelectedLang = _lang.code === lang.code;
              return (
                <Button
                  other={{ disabled: isSelectedLang }}
                  ariaLabel={_lang.label}
                  onClick={() => setLang(_lang)}
                  key={_lang.code}
                  className={classnames(styles.lang, { [styles.selectedLang]: isSelectedLang })}
                >
                  <LangFlagMap langCode={_lang.code} />
                </Button>
              );
            })}
          </div>
        ) : null
      }

      <Button
        onClick={() => setIsMenuOpen(!isMenuOpen)}
        className={styles.menuButton}
        ariaLabel={isMenuOpen ? t('Close Menu') : t('Open Menu')}
      >
        {isMenuOpen ? <ArrowDownIcon className={styles.arrowIcon} /> : <ArrowUpIcon className={styles.arrowIcon} />}
      </Button>
      <div className={styles.menuContent}>
        <div className={styles.content}>
          {h1Title}
          {gameDescription}
          {homeButton}
        </div>
      </div>
    </div>
  );
};

const Menu = compose(
  connect(
    createStructuredSelector<
      IState,
      {
        isMenuOpen: IProps['isMenuOpen'];
        selectedGame: IProps['selectedGame'];
        availableLangs: IProps['availableLangs'];
        lang: IProps['lang'];
      },
      {
        isMenuOpen: IProps['isMenuOpen'];
        selectedGame: IProps['selectedGame'];
        availableLangs: IProps['availableLangs'];
        lang: IProps['lang'];
      }
    >({
      isMenuOpen,
      selectedGame,
      availableLangs,
      lang
    }),
    { setIsMenuOpen, setLang }
  ),
  withTranslation
)(_Menu);

export { Menu };
