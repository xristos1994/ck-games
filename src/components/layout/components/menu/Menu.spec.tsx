import React, { MouseEventHandler, ReactChild } from 'react';
import renderer from 'react-test-renderer';
import { _Menu as Menu } from './Menu';

const Button = (
  { children, onClick, className='' } :
  { children: ReactChild | ReactChild[]; onClick: MouseEventHandler<HTMLButtonElement>; className: string }
) => {
  return (
    <button onClick={onClick} className={className || undefined}>
      {children}
    </button>
  );
};

jest.mock('@components', () => ({
  __esModule: true,
  Button: Button
}));

const ArrowUpIcon = () => {
  return (
    <div>
      ArrowUpIcon
    </div>
  );
};

const ArrowDownIcon = () => {
  return (
    <div>
      ArrowDownIcon
    </div>
  );
};

jest.mock('@components/icons', () => ({
  __esModule: true,
  ArrowUpIcon: ArrowUpIcon,
  ArrowDownIcon: ArrowDownIcon
}));

const AvailableGames = {
  pantomime: 'pantomime',
  tikTakBoom: 'tik-tak-boom'
};

jest.mock('@models/website/interfaces', () => ({
  __esModule: true,
  AvailableGames: AvailableGames
}));

const t = jest.fn().mockImplementation((label: string, param?: string[]) => label + (param ? param[0] : ''));
const setIsMenuOpen = jest.fn();
const setLang = jest.fn();
const availableLangs = [
  { code: 'EN', label: 'English' },
  { code: 'EL', label: 'Ελληνικά' }
];

// selectedGame: string | null;
// isMenuOpen: boolean;
// setIsMenuOpen: (isMenuOpen: boolean) => void;
// availableLangs: { code: string; label: string }[];
// lang: { code: string; label: string };
// setLang: (lang: IProps['lang']) => void;
// t: ITranslate;

describe('components/Menu', () => {
  it('renders correctly', () => {
    const initialWindow = { ...global.window };
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    global.window = undefined;

    let tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.pantomime}
        isMenuOpen={false}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.tikTakBoom}
        isMenuOpen={false}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Menu
        selectedGame={null}
        isMenuOpen={false}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    global.window = {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      location: {
        pathname: ''
      }
    };

    tree = renderer.create(
      <Menu
        selectedGame={null}
        isMenuOpen={true}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.pantomime}
        isMenuOpen={true}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.tikTakBoom}
        isMenuOpen={true}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.pantomime}
        isMenuOpen={false}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.tikTakBoom}
        isMenuOpen={false}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    global.window = { ...initialWindow };
  });

  it('has the right event handlers', () => {
    const initialWindow = { ...global.window };
    global.window = {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      location: {
        pathname: ''
      }
    };

    const tree = renderer.create(
      <Menu
        selectedGame={AvailableGames.pantomime}
        isMenuOpen={true}
        setIsMenuOpen={setIsMenuOpen}
        availableLangs={availableLangs}
        lang={availableLangs[0]}
        setLang={setLang}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    );

    const buttons = tree.root.findAllByType('button');
    buttons[1].props.onClick();
    expect(setLang).toHaveBeenLastCalledWith(availableLangs[1]);

    buttons[2].props.onClick();
    expect(setIsMenuOpen).toHaveBeenLastCalledWith(false);

    global.window = { ...initialWindow };
  });
});
