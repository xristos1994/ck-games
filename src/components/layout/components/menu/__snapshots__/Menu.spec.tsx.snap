// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Menu renders correctly 1`] = `
<div
  className="menuContainer"
>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowUpIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      >
        Pantomime
      </h1>
      <div
        className="gameDescription"
      >
        <h2>
          Tik-Tak-Boom
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Tik-Tak-Boom Description HTML",
          }
        }
      />
      <div
        className="gameDescription"
      >
        <h2>
          Pantomime
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Pantomime Description HTML",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 2`] = `
<div
  className="menuContainer"
>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowUpIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      >
        Tik-Tak-Boom
      </h1>
      <div
        className="gameDescription"
      >
        <h2>
          Tik-Tak-Boom
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Tik-Tak-Boom Description HTML",
          }
        }
      />
      <div
        className="gameDescription"
      >
        <h2>
          Pantomime
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Pantomime Description HTML",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 3`] = `
<div
  className="menuContainer"
>
  <div
    className="langSelector"
  >
    <button
      className="lang selectedLang"
      onClick={[Function]}
    >
      <span>
        EN
      </span>
    </button>
    <button
      className="lang"
      onClick={[Function]}
    >
      <span>
        EL
      </span>
    </button>
  </div>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowUpIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      >
        Menu Title SSR
      </h1>
      <div
        className="gameDescription"
      >
        <h2>
          Tik-Tak-Boom
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Tik-Tak-Boom Description HTML",
          }
        }
      />
      <div
        className="gameDescription"
      >
        <h2>
          Pantomime
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Pantomime Description HTML",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 4`] = `
<div
  className="menuContainer openMenu"
>
  <div
    className="langSelector"
  >
    <button
      className="lang selectedLang"
      onClick={[Function]}
    >
      <span>
        EN
      </span>
    </button>
    <button
      className="lang"
      onClick={[Function]}
    >
      <span>
        EL
      </span>
    </button>
  </div>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowDownIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      >
        CK-Games
      </h1>
      <div
        className="gameDescription"
      >
        <h2>
          Tik-Tak-Boom
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Tik-Tak-Boom Description HTML",
          }
        }
      />
      <hr />
      <div
        className="gameDescription"
      >
        <h2>
          Pantomime
        </h2>
      </div>
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Pantomime Description HTML",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 5`] = `
<div
  className="menuContainer openMenu"
>
  <div
    className="langSelector"
  >
    <button
      className="lang selectedLang"
      onClick={[Function]}
    >
      <span>
        EN
      </span>
    </button>
    <button
      className="lang"
      onClick={[Function]}
    >
      <span>
        EL
      </span>
    </button>
  </div>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowDownIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      >
        Pantomime
      </h1>
      <div
        className="gameDescription"
      />
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Pantomime Description HTML",
          }
        }
      />
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        <button
          className="homeButton"
        >
          Home
        </button>
      </a>
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 6`] = `
<div
  className="menuContainer openMenu"
>
  <div
    className="langSelector"
  >
    <button
      className="lang selectedLang"
      onClick={[Function]}
    >
      <span>
        EN
      </span>
    </button>
    <button
      className="lang"
      onClick={[Function]}
    >
      <span>
        EL
      </span>
    </button>
  </div>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowDownIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      >
        Tik-Tak-Boom
      </h1>
      <div
        className="gameDescription"
      />
      <div
        className="gameDescription"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Tik-Tak-Boom Description HTML",
          }
        }
      />
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        <button
          className="homeButton"
        >
          Home
        </button>
      </a>
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 7`] = `
<div
  className="menuContainer"
>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowUpIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      />
    </div>
  </div>
</div>
`;

exports[`components/Menu renders correctly 8`] = `
<div
  className="menuContainer"
>
  <button
    className="menuButton"
    onClick={[Function]}
  >
    <div>
      ArrowUpIcon
    </div>
  </button>
  <div
    className="menuContent"
  >
    <div
      className="content"
    >
      <h1
        className="menuTitle"
      />
    </div>
  </div>
</div>
`;
